File: src/main/java/com/adamkali/simpleide/editor/io/Document.java
package com.adamkali.simpleide.editor.io;

import com.adamkali.simpleide.editor.lang.Lexer;
import com.adamkali.simpleide.editor.lang.tokens.Token;
import com.adamkali.simpleide.editor.lang.tokens.character.NewLineToken;

import java.util.ArrayList;
import java.util.List;

public class Document {
    private List<Line> lines;

    public Document() {
        lines = new ArrayList<>() {{
            add(new Line());
        }};
//        lines = new ArrayList<>() {{
//            add(new Line() {{ addToken(new NewLineToken()); }});
//            add(new Line() {{ addToken(new SLCommentToken("// This is a comment")); addToken(new NewLineToken());  }});
//            add(new Line() {{ addToken(new KeywordToken("for")); addToken(new WhitespaceToken()); addToken(new UndefinedToken("test")); }});
//        }};
    }

    public List<Line> getLines() {
        return lines;
    }

    /**
     * Inserts a new line after the given line.
     *
     * @param lineBefore The line to insert the new line after.
     */
    public void newLine(int lineBefore) {
        if (lineBefore < 0 || lineBefore >= numberOfLines()) {
            throw new IndexOutOfBoundsException("Line index out of bounds");
        }

        lines.add(lineBefore + 1, new Line());
    }

    public Line getLine(int line) {
        if (line < 0 || line >= numberOfLines()) {
            throw new IndexOutOfBoundsException("Line index out of bounds");
        }
        return lines.get(line);
    }

    public void removeToken(int line, int index) {
        if (line < 0 || line >= numberOfLines()) {
            throw new IndexOutOfBoundsException("Line index out of bounds");
        }
        lines.get(line).removeToken(index);
    }

    /**
     * Removes a line from the document.
     *
     * @param line Line number
     */
    public void removeLine(int line) {
        if (line < 0 || line >= numberOfLines()) {
            throw new IndexOutOfBoundsException("Line index out of bounds");
        }
        lines.remove(line);
    }

    public int numberOfLines() {
        return lines.size();
    }

    public static class Line {
        private List<Token> tokens;

        public Line() {
            tokens = new ArrayList<>();
        }

        public List<Token> getTokens() {
            return tokens;
        }

        public void rewrite(String newText) {
            List<Token> newTokens = Lexer.lex(newText);
            // Check there's no newline in the middle of the line
            for (int index = 0; index < newTokens.size(); index++) {
                if (newTokens.get(index) instanceof NewLineToken && index != newTokens.size() - 1) {
                    throw new IllegalArgumentException("Newline in the middle of the line");
                }
            }
            tokens = newTokens;
        }

        public void append(String textToAppend) {
            rewrite(this + textToAppend);
        }

        public void insert(int index, String textToInsert) {
            rewrite(toString().substring(0, index) + textToInsert + toString().substring(index));
        }

        public void addToken(Token token) {
            tokens.add(token);
        }

        public void removeToken(int index) {
            tokens.remove(index);
        }

        /**
         * Returns the string length of the line.
         *
         * @return
         */
        public int length() {
            int length = 0;
            for (Token token : tokens) {
                length += token.getText().length();
            }
            return length;
        }

        public String toString() {
            StringBuilder sb = new StringBuilder();
            for (Token token : tokens) {
                sb.append(token.getText());
            }
            return sb.toString();
        }

        public String substring(int start, int end) {
            if (start < 0 || end < 0 || start > end || end > length()) {
                throw new IndexOutOfBoundsException("Index out of bounds");
            }
            return toString().substring(start, end);
        }
    }
}

File: src/main/java/com/adamkali/simpleide/editor/io/Document.kt
package com.adamkali.simpleide.editor.io


class Document {
    var lines: MutableList<Line> = mutableListOf()
        private set

    /**
     * Whether [toText] should end with a newline. Loaded from the original file
     * so a trailing newline is not painted as an extra empty line.
     */
    var trailingNewline: Boolean = false

    init {
        lines.add(Line())
    }

    /**
     * Replaces the document with [text], normalizing CR/CRLF to LF.
     * An empty file becomes one empty line. A trailing newline is stored in
     * [trailingNewline] rather than as an extra painted line.
     */
    fun replaceText(text: String) {
        val normalized = text.replace("\r\n", "\n").replace("\r", "\n")
        trailingNewline = normalized.endsWith("\n")
        val body = if (trailingNewline) normalized.removeSuffix("\n") else normalized
        val parts = if (body.isEmpty()) {
            listOf("")
        } else {
            body.split("\n")
        }

        lines.clear()
        for (part in parts) {
            val line = Line()
            line.rewrite(part)
            lines.add(line)
        }
        if (lines.isEmpty()) {
            lines.add(Line())
        }
    }

    /**
     * Serializes the document to a single string. Lines are joined with `\n`.
     * When [trailingNewline] is set, the result ends with `\n`.
     */
    fun toText(): String {
        val body = lines.joinToString("\n") { it.toString() }
        return if (trailingNewline) "$body\n" else body
    }

    fun insertLine(lineIndex: Int, line: Line) {
        if (lineIndex < 0 || lineIndex > lines.size) {
            throw IndexOutOfBoundsException("Index out of bounds (lineIndex: $lineIndex)")
        }
        lines.add(lineIndex, line)
    }

    /**
     * Adds a new line after a line.
     * @param lineIndex The index of the line to add a new line after.
     */
    fun newLine(lineIndex: Int) {
        if (lineIndex < 0 || lineIndex > lines.size) {
            throw IndexOutOfBoundsException("Index out of bounds (lineIndex: $lineIndex)")
        }
        lines.add(lineIndex + 1, Line())
    }

    /**
     * Gets a line.
     * @param lineIndex The index of the line to get.
     */
    fun getLine(lineIndex: Int): Line {
        if (lineIndex < 0 || lineIndex >= lines.size) {
            throw IndexOutOfBoundsException("Index out of bounds (lineIndex: $lineIndex)")
        }
        return lines[lineIndex]
    }

    /**
     * Removes a token from a line
     * @param line The index of the line to remove the token from.
     * @param index The index of the token to remove.
     */
    fun removeToken(line: Int, index: Int) {
        if (line < 0 || line >= lines.size) {
            throw java.lang.IndexOutOfBoundsException("Line index out of bounds")
        }
        lines[line].removeToken(index)
    }

    /**
     * Removes a line from the document.
     * @param line Line number
     */
    fun removeLine(line: Int) {
        if (line < 0 || line >= lines.size) {
            throw java.lang.IndexOutOfBoundsException("Line index out of bounds")
        }
        lines.removeAt(line)
    }

    /**
     * Swaps the lines at [a] and [b].
     */
    fun swapLines(a: Int, b: Int) {
        if (a < 0 || a >= lines.size || b < 0 || b >= lines.size) {
            throw IndexOutOfBoundsException("Index out of bounds (a: $a, b: $b)")
        }
        val first = lines[a]
        lines[a] = lines[b]
        lines[b] = first
    }

    /**
     * Returns the number of lines in the document.
     * @return The number of lines in the document.
     */
    fun getLineCount(): Int {
        return lines.size
    }


}